Zend Coding Standard 2

xtreamwayz opened the pull request:
This PR adds a stricter ruleset. #1 is adding a lot of extra code and custom rules which are hard to maintain and it's so huge that it is hard to review. The huge amount of work done in #1 is not lost and is moved to [webimpress/coding-standard](https://github.com/webimpress/coding-standard).

The purpose of this PR is to start over and importing rules from other packages where possible.

- [x] Create new structure
- [x] Make the ruleset auto discoverable so it can be used globally with `phpcs -s --standard=ZendCodingStandard src` 
- [x] Import [slevomat/coding-standard rules](https://github.com/slevomat/coding-standard/)
- [x] _**REVIEW PHASE**_: The rules need to be reviewed before the documentation is being written
- [x] Remove rules that are too strict for zendframework
- [x] Add tests
- [x] Write documentation for https://docs.zendframework.com/zend-coding-style/
- [x] Document deprecated changes: [ignore parts](https://github.com/squizlabs/PHP_CodeSniffer/wiki/Advanced-Usage#ignoring-parts-of-a-file) in 4.0
- [x] Add a `CHANGELOG.md` entry
- [ ] Test code style on existing packages
  - https://github.com/zendframework/zend-diactoros/pull/331

# Blocked by:
- [x] ~~Poll: [To align or not to align!](https://discourse.zendframework.com/t/poll-zend-coding-standard-2-aligning-statements-arrays-and-comment-tags/729)~~
- [x] ~~https://github.com/slevomat/coding-standard/issues/477~~ fixed in 4.8.0
- [x] ~~https://github.com/slevomat/coding-standard/issues/478~~ fixed in 4.8.0
- [x] ~~Aligning double arrows in arrays~~

# Testing
Make sure you remove the global installation after testing from your global composer.json file!!!
```bash
$ composer global config repositories.zend-coding-standard vcs git@github.com:xtreamwayz/zend-coding-standard.git
$ composer global require --dev zendframework/zend-coding-standard:dev-feature/new-rules

# For this to work, add this to your path: ~/.composer/vendor/bin
# Using `-s` prints the rules that triggered the errors so they can be reviewed easily
$ phpcs -sp --standard=ZendCodingStandard src

# Automatically fixing errors
$ phpcbf -sp --standard=ZendCodingStandard src
```

weierophinney commented on src/ZendCodingStandard/ruleset.xml:
```diff
@@ -0,0 +1,379 @@
+                    is_double => is_float,
+                    is_integer => is_int,
+                    is_long => is_int,
+                    is_null => null,
```
What happens when this fix is applied? Seems like it might break code, unless `null` here means `null ===`?

xtreamwayz replied:
Need to look into this what it exactly does, but I think it is replaced with `null ===`.

xtreamwayz replied:
It does not replace the functions. It only displays an error:
```
 19 | ERROR | The use of function chop() is forbidden; use rtrim()
    |       | instead
 20 | ERROR | The use of function sizeof() is forbidden; use count()
    |       | instead
 21 | ERROR | The use of function is_null() is forbidden
 23 | ERROR | The use of function settype() is forbidden
 30 | ERROR | The use of function extract() is forbidden
 31 | ERROR | The use of function compact() is forbidden
```

xtreamwayz replied:
null means there is no alternative. `is_null` should be replaced with `null ===` but changing this to that value gives the following error:
```
21 | ERROR | The use of function is_null() is forbidden; use 'null
   |       | ==='() instead
```
That looks like a weird function or so.

weierophinney commented on src/ZendCodingStandard/ruleset.xml:
```diff
@@ -0,0 +1,379 @@
+    <!-- Forbid backtick operator -->
+    <rule ref="Generic.PHP.BacktickOperator"/>
+    <!-- Forbid short open tag -->
+    <rule ref="Generic.PHP.DisallowShortOpenTag"/>
```
Does this disallow `<?=`? I ask, as this is valid _always_ starting in 5.4 or 5.5, and we recommend it for templates. If it doesn't, we're good; if it does, is there a way to make an exception for `<?=`?

xtreamwayz replied:
I'm not sure if this can be used to validate templates. I've never tried it. But yes, if phpcs works with templates then it should be removed.

weierophinney replied:
The question is whether or not it flags `<?=` as invalid. If it doesn't, we're fine. :)

weierophinney commented on src/ZendCodingStandard/ruleset.xml:
```diff
@@ -0,0 +1,379 @@
+                    @category,
+                    @created,
+                    @package,
+                    @since,
```
Why is this one considered forbidden? It's useful for noting API additions...

xtreamwayz replied:
I think the reasoning behind it is that all info is in git log and the changelog. But I'll remove it since it's easy to look it up directly in the code where most developers look anyway.

weierophinney commented on src/ZendCodingStandard/ruleset.xml:
```diff
@@ -0,0 +1,379 @@
+    <!-- Require comments with single line written as one-liners -->
+    <rule ref="SlevomatCodingStandard.Commenting.RequireOneLinePropertyDocComment"/>
+    <!-- Forbid fancy yoda conditions -->
+    <rule ref="SlevomatCodingStandard.ControlStructures.DisallowYodaComparison"/>
```
Disagree with this one; we use `(null|false|true) [!=]== <expression>` all over the place as a defensive programming approach.

weierophinney commented on src/ZendCodingStandard/ruleset.xml:
```diff
@@ -0,0 +1,379 @@
+    <!-- Require the usage of assignment operators, eg `+=`, `.=` when possible -->
+    <rule ref="SlevomatCodingStandard.Operators.RequireCombinedAssignmentOperator"/>
+    <!-- Forbid `list(...)` syntax, use [...] instead -->
+    <rule ref="SlevomatCodingStandard.PHP.ShortList"/>
```
This one will only work for libraries on PHP 7+; for those on 5.5, we'd need to disable this rule.

xtreamwayz replied:
The idea is to tag this release 2.0 and include it `^7.1` packages. I don't see why we need to fix code for old packages. That's a lot of extra work and commits.

weierophinney replied:
Fair enough!

weierophinney commented on src/ZendCodingStandard/ruleset.xml:
```diff
@@ -0,0 +1,379 @@
+        </properties>
+    </rule>
+    <!-- Forbid useless @var for constants -->
+    <rule ref="SlevomatCodingStandard.TypeHints.UselessConstantTypeHint"/>
```
I disagree with this one. The proposed [phpdoc spec specifically states @var is the appropriate tag for use with constants](https://github.com/php-fig/fig-standards/blob/master/proposed/phpdoc-tags.md#522-var).

weierophinney commented on src/ZendCodingStandard/ruleset.xml:
```diff
@@ -0,0 +1,379 @@
+    <!-- Forbid `global` -->
+    <rule ref="Squiz.PHP.GlobalKeyword"/>
+    <!-- Forbid functions inside functions -->
+    <rule ref="Squiz.PHP.InnerFunctions"/>
```
Why? This is a valid way to make private functions...

froschdesign commented on mkdocs.yml:
```diff
@@ -0,0 +1,17 @@
+site_name: Zend Coding Standard
+site_description: 'The coding standard ruleset for Zend Framework components.'
+repo_url: 'https://github.com/zendframework/zend-coding-standard'
+copyright: 'Copyright (c) 2016-2018 <a href="https://www.zend.com/">Zend Technologies USA Inc.</a>'
```
This line can be removed, because this is not used in theme.

froschdesign commented on docs/book/index.html:
```diff
@@ -0,0 +1,9 @@
+<div class="container">
```
Remove the entire file, the same content is generated by the theme.

froschdesign commented on .travis.yml:
```diff
@@ -1,55 +1,37 @@
+      apt:
+        packages:
+        - libxml2-utils
+    script: xmllint --noout --schema vendor/squizlabs/php_codesniffer/phpcs.xsd src/ZendCodingStandard/ruleset.xml
```
Very nice! 👍

xtreamwayz commented:
@mwop @froschdesign I've fixed the issues, added tests and I'm now working on the docs.
70 files is a lot, but it's still only 1 ruleset file and the rest is tests and documentation.

xtreamwayz commented:
All documentation is added. I've left plenty of typo's, inconsistencies and mistakes for others to
discover. Some rules in the ruleset.md file need examples.

/cc @mwop, @froschdesign

xtreamwayz commented on src/ZendCodingStandard/ruleset.xml:
```diff
@@ -0,0 +1,389 @@
+    <!-- Force `self::` for self-reference, force lower-case self, forbid spaces around `::` -->
+    <rule ref="Squiz.Classes.SelfMemberReference"/>
+    <!-- Force phpDoc alignment -->
+    <rule ref="Squiz.Commenting.DocCommentAlignment">
```
Before:
```
    /**
     * Create a plain text response.
     *
     * Produces a text response with a Content-Type of text/plain and a default
     * status of 200.
     *
     * @param string|StreamInterface $text String or stream for the message body.
     * @param int $status Integer status code for the response; 200 by default.
     * @param array $headers Array of headers to use at initialization.
     * @throws Exception\InvalidArgumentException if $text is neither a string or stream.
     */
```
After:
```
    /**
     * Create a plain text response.
     *
     * Produces a text response with a Content-Type of text/plain and a default
     * status of 200.
     *
     * @param string|StreamInterface $text    String or stream for the message body.
     * @param int                    $status  Integer status code for the response; 200 by default.
     * @param array                  $headers Array of headers to use at initialization.
     * @throws Exception\InvalidArgumentException if $text is neither a string or stream.
     */
```

xtreamwayz replied:
Before:
```
    /**
     * @param array $serverParams Server parameters, typically from $_SERVER
     * @param array $uploadedFiles Upload file information, a tree of UploadedFiles
     * @param null|string|UriInterface $uri URI for the request, if any.
     * @param null|string $method HTTP method for the request, if any.
     * @param string|resource|StreamInterface $body Message body, if any.
     * @param array $headers Headers for the message, if any.
     * @param array $cookies Cookies for the message, if any.
     * @param array $queryParams Query params for the message, if any.
     * @param null|array|object $parsedBody The deserialized body parameters, if any.
     * @param string $protocol HTTP protocol version.
     * @throws Exception\InvalidArgumentException for any invalid value.
     */
```
After:
```
    /**
     * @param array                           $serverParams  Server parameters, typically from $_SERVER
     * @param array                           $uploadedFiles Upload file information, a tree of UploadedFiles
     * @param null|string|UriInterface        $uri           URI for the request, if any.
     * @param null|string                     $method        HTTP method for the request, if any.
     * @param string|resource|StreamInterface $body          Message body, if any.
     * @param array                           $headers       Headers for the message, if any.
     * @param array                           $cookies       Cookies for the message, if any.
     * @param array                           $queryParams   Query params for the message, if any.
     * @param null|array|object               $parsedBody    The deserialized body parameters, if any.
     * @param string                          $protocol      HTTP protocol version.
     * @throws Exception\InvalidArgumentException for any invalid value.
     */
```

weierophinney replied:
Let's remove the DocCommentAlignment rule.

I have been moving away from alignment in docblock annotations for some time now as they introduce unnecessary noise in diffs, and because doing so often force the contents past a reasonable line length. Additionally, when you split such comments over multiple lines, doing so breaks the alignment.

I'd much rather we don't align, and have the second and later lines indented.

Ocramius replied:
Same as above: all tooling around git can ignore whitespace, while I really can't read this (taken from @xtreamwayz' comment)

![screenshot_20180911-092922](https://user-images.githubusercontent.com/154256/45345096-9d69fb80-b5a5-11e8-8805-d9554b4117f5.png)

xtreamwayz commented on src/ZendCodingStandard/ruleset.xml:
```diff
@@ -0,0 +1,389 @@
+    <!-- report invalid format of inline phpDocs with @var -->
+    <rule ref="SlevomatCodingStandard.Commenting.InlineDocCommentDeclaration"/>
+    <!-- Require comments with single line written as one-liners -->
+    <rule ref="SlevomatCodingStandard.Commenting.RequireOneLinePropertyDocComment"/>
```
Before:
```
    /**
     * @var callable|null
     */
    protected $callback;
```
After:
```
    /** @var callable|null */
    protected $callback;
```

weierophinney replied:
What does this do when a multi-line property comment is provided? For instance, one that has an actual description? If those are left alone, and it only reformats ones such as these, I'd be fine with the rule. Otherwise, we need to remove it.

xtreamwayz replied:
I've checked the sniff and it only affects one-line comments with `@var` in it.

```
    /** @var int[] Array indexed by valid scheme names to their corresponding ports. */
    protected $allowedSchemes = [];

    /** @var string */
    private $scheme = '';

    /**
     * Is a given port non-standard for the current scheme?
     */
    private function isNonStandardPort(string $scheme, string $host, ?int $port) : bool
    {
        // ...
    }

    /**
     * {@inheritdoc}
     */
    public function getHost() : string
    {
        // ...
    }
```

xtreamwayz commented on src/ZendCodingStandard/ruleset.xml:
```diff
@@ -0,0 +1,389 @@
+    <!-- Forbid weak comparisons -->
+    <rule ref="SlevomatCodingStandard.ControlStructures.DisallowEqualOperators"/>
+    <!-- Require usage of early exit -->
+    <rule ref="SlevomatCodingStandard.ControlStructures.EarlyExit"/>
```
Before:
```
        // per PSR-7: attempt to set the Host header from a provided URI if no
        // Host header is provided
        if (! $this->hasHeader('Host') && $this->uri->getHost()) {
            $this->headerNames['host'] = 'Host';
            $this->headers['Host'] = [$this->getHostFromUri()];
        }
```
After:
```
        // per PSR-7: attempt to set the Host header from a provided URI if no
        // Host header is provided
        if ($this->hasHeader('Host') || ! $this->uri->getHost()) {
            return;
        }

        $this->headerNames['host'] = 'Host';
        $this->headers['Host']     = [$this->getHostFromUri()];
```

xtreamwayz replied:
Before:
```
        // Remove an existing host header if present, regardless of current
        // de-normalization of the header name.
        // @see https://github.com/zendframework/zend-diactoros/issues/91
        foreach (array_keys($new->headers) as $header) {
            if (strtolower($header) === 'host') {
                unset($new->headers[$header]);
            }
        }
```
After:
```
        // Remove an existing host header if present, regardless of current
        // de-normalization of the header name.
        // @see https://github.com/zendframework/zend-diactoros/issues/91
        foreach (array_keys($new->headers) as $header) {
            if (strtolower($header) !== 'host') {
                continue;
            }

            unset($new->headers[$header]);
        }
```

xtreamwayz replied:
Before:
```
    private function parseUri(string $uri) : void
    {
        // ...

        if (isset($parts['pass'])) {
            $this->userInfo .= ':' . $parts['pass'];
        }
    }
```
After:
```
    private function parseUri(string $uri) : void
    {
        // ...

        if (! isset($parts['pass'])) {
            return;
        }

        $this->userInfo .= ':' . $parts['pass'];
    }
```

weierophinney replied:
I think this rule is overly aggressive in what it does when "fixing" the code. Let's drop it for now.

xtreamwayz commented on src/ZendCodingStandard/ruleset.xml:
```diff
@@ -0,0 +1,389 @@
+        <exclude name="Squiz.Arrays.ArrayDeclaration.ValueNotAligned"/>
+        <!-- Checked by SlevomatCodingStandard.Arrays.TrailingArrayComma.MissingTrailingComma -->
+        <exclude name="Squiz.Arrays.ArrayDeclaration.NoCommaAfterLast"/>
+    </rule>
```
This array rule needs some tweaking.

Current situation:
```
$serializedRequest = [
    'method' => 'POST',
    'request_target' => '/foo/bar?baz=bat',
    'uri' => 'http://example.com/foo/bar?baz=bat',
    'protocol_version' => '1.1',
    'headers' => [
        'Host' => ['example.com'],
        'Accept' => ['application/json'],
        'X-Foo-Bar' => [
            'Baz',
            'Bat',
        ],
    ],
    'body' => '{"test":"value"}',
];
```

In several components the double arrows are aligned. The question is should those be aligned or not? The reason for doing it is readability. On the other hand, PHP supports the trailing comma on the last array element for a reason: To remove clutter from commit diffs. Aligning the double arrows brings back what the trailing comma prevents.

With double arrow indentation:
```
$serializedRequest = [
    'method'           => 'POST',
    'request_target'   => '/foo/bar?baz=bat',
    'uri'              => 'http://example.com/foo/bar?baz=bat',
    'protocol_version' => '1.1',
    'headers'          => [
        'Host'      => ['example.com'],
        'Accept'    => ['application/json'],
        'X-Foo-Bar' => [
            'Baz',
            'Bat',
        ],
    ],
    'body'             => '{"test":"value"}',
];
```

weierophinney replied:
I'm okay for removing alignment generally. The readability benefits are typically not huge, and enforcing it means a lot of extra changes for what are often trivial additions.

Ocramius replied:
Disagree here: readability is king, and `git diff`, `git blame/annotate` as well as the github and gitlab UI have flags for ignoring whitespace changes.

xtreamwayz replied:
@Ocramius you have a good point there. I totally forgot about the flag to ignore whitespace changes.

weierophinney commented on composer.json:
```diff
@@ -1,12 +1,51 @@
 {
     "name": "zendframework/zend-coding-standard",
-    "description": "Zend Framework coding standard",
+    "type": "phpcodesniffer-standard",
```
I'm so happy that composer has made things like plugin registration automated via package `type`!

xtreamwayz commented:
Merged into develop
